americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+        yticklabels[0] = 0
+        ax.set_yticklabels(yticklabels, size=kwargs['ytick_size'])
+
+    # Adds a legend to the top to distinguish groups
+    # Determines the horizontal ordering for the groups
+    num_groups = len(order)
+    num_leg_cols = int(np.ceil(float(num_groups)/2))
+    leg_order = np.hstack([np.array([i, i+num_leg_cols])
+                           for i in xrange(num_leg_cols)])[:num_groups]
+    # Gets the patches and labels
+    patches = np.array(ax.patches[:num_groups])[leg_order]
+    labels = np.array([l.split('(')[0].replace('_', ' ')
+                       for l in order])[leg_order]
+    # Adds the legend
+    ax.legend(patches, labels,
+              loc="upper center",
+              ncol=num_leg_cols,
+              fontsize=12,
+              frameon=False)
+
+    return ax
+
+
+def split_taxa(taxa, level=7):
+    """Splits a greengenes taxonomy string into a dataframe
+
+    Parameters
+    ----------
+    taxa : list
+        a list of greengenes strings
+    level: {2, 3, 4, 5, 6, 7}
+        the taxonomic level for summary
+
+    Returns
+    -------
+    splits : array
+        the taxonomy data summarized to the specified level. If data is missing
+        for a particular level, the string will specify the last known string.
+        (i.e. k__kingdom;p__phylum;c__pclass;o__porder;f__family;g__;s__
+        outputs with a genus and species value "f. family").
+
+    levels: list
+        the name of the taxonomic levels corresponding to columns in the table.
+    """
+
+    if level not in {2, 3, 4, 5, 6, 7}:
+        raise ValueError('%r is not a supported taxonomic level.' % level)
+
+    # Prealocates a holding object
+    splits = np.zeros((1, (level)))
+
+    # Sets up the names of phylogenetic levels
+    levels = ['kingdom', 'phylum', 'p_class', 'p_order', 'family', 'genus',
+              'species']
+
+    # Loops through the list of taxa
+    for taxon in taxa:
+        if taxon == 'Unassigned':
+            clean = ['Unassigned']*level
+        else:
+            # Splits the greengenes string
+            rough = [v.strip() for v in taxon.split(';')]
+            # Watches the last space in the data
+            try:
+                last = rough[level - 1].replace('__', '. ')
+            except:
+                raise ValueError('There are %i levels in your taxa string and'
+                                 ' you would like to look at %i levels.'
+                                 % (len(rough), level))
+            clean = []
+
+            # Cleans the taxa
+            for t in rough[:(level)]:
+                val = t.split('__')[1].replace('_', ' ').replace('-', ' ')
+                if val == '':
+                    clean.append(last)
+                elif '[' in val:
+                    v_corr = 'cont. %s' % val.replace('[', '').replace(']', '')
+                    clean.append(v_corr)
+                    last = '%s. %s' % (t.split('__')[0], v_corr)
```

Comment:
only need to do the split on `__` once. suggest setting up the rank, taxon name split first thing in the for loop